### PR TITLE
Reverse playback support

### PR DESCRIPTION
**Pull request by cgomezv**

### CODE REVIEW

File: omx/gstomx.c [hunk 1]
```diff
@@ -1251,13 +1248,14 @@ gst_omx_port_compare_buffers (gconstpointer a, gconstpointer b)
  * If *buf is not NULL the buffer containing the given omx pointer
  * is going to be returned */
 GstOMXAcquireBufferReturn
-gst_omx_port_acquire_buffer (GstOMXPort * port, GstOMXBuffer ** buf)
+gst_omx_port_acquire_buffer (GstOMXPort * port, GstOMXBuffer ** buf, gboolean retry_limit)
```
Comment (joselopez10) [inline omx/gstomx.c:1257]: Add retry_limit explanation/comments somewhere in the function.

File: omx/gstomx.c [hunk 2]
```diff
@@ -1277,6 +1275,11 @@ gst_omx_port_acquire_buffer (GstOMXPort * port, GstOMXBuffer ** buf)
   if (timeout != 40 * GST_MSECOND )
     timeout = -2;
 
+  if (retry_limit && num_retries > 5) {
```
Comment (joselopez10) [inline omx/gstomx.c]: Consider using something like GST_OMX_NUM_RETRIES_DEFAULT 5, same for 40 * GST_MSECONDS

File: omx/gstomx.c [hunk 3]
```diff
@@ -1384,6 +1387,7 @@ gst_omx_port_acquire_buffer (GstOMXPort * port, GstOMXBuffer ** buf)
     gst_omx_component_wait_message (comp, timeout == -2 ? GST_SECOND : timeout);
 
     /* And now check everything again and maybe get a buffer */
+    num_retries++;
```
Comment (joselopez10) [inline omx/gstomx.c:1396]: Having an extra attribute inside GstOmxPort an option to determine if we have to use the num_retries or not? That could prevent changing the acquire_buffer signature.

**Review by joselopez10 — CHANGES_REQUESTED** [review]
Check inline  comments.

File: omx/gstomx.c (hunk 1 above)
Comment (cgomezv) [reply omx/gstomx.c:1257]: Added in commit 577747aa1d7afe875676172b923ad778d8ccd06b

File: omx/gstomx.c (hunk 2 above)
Comment (cgomezv) [reply omx/gstomx.c]: Changed in commit 577747aa1d7afe875676172b923ad778d8ccd06b

File: omx/gstomx.c (hunk 3 above)
Comment (cgomezv) [reply omx/gstomx.c:1396]: It seems that GstOmxPort structure is more intended to exchange information about the component port settings, a change in there would spread through all the set_format functions of many elements available in this repository. The retry_limit is more a feature of the acquire_buffer  function itself, it modifies the internal flow of the acquire buffer logic as required.

File: omx/gstomxvideodec.c [hunk 4]
```diff
@@ -1290,33 +1290,30 @@ gst_omx_video_dec_loop (GstOMXVideoDec * self)
...
 #else
   port = self->dec_out_port;
 #endif
 
-  GST_VIDEO_DECODER_STREAM_LOCK (self);
-  flush_flag = self->flush_flag;
-  GST_VIDEO_DECODER_STREAM_UNLOCK (self);
-
-  if (flush_flag) {
+  if (self->flush_flag) {
```
Comment (mmontero) [inline omx/gstomxvideodec.c]: keep flush flag protection

File: omx/gstomxvideodec.c [hunk 5]
```diff
@@ -1620,7 +1617,7 @@ gst_omx_video_dec_loop (GstOMXVideoDec * self)
     self->downstream_flow_ret = flow_ret;
 
     /* Here we fallback and pause the task for the EOS case */
-    if (flow_ret != GST_FLOW_OK)
+    if (flow_ret != GST_FLOW_OK && !self->flush_flag)
```
Comment (mmontero) [inline omx/gstomxvideodec.c:1628]: protect flush_flag read

File: omx/gstomxvideodec.c [hunk 6]
```diff
@@ -2354,18 +2351,13 @@ gst_omx_video_dec_handle_frame (GstVideoDecoder * decoder,
...
   klass = GST_OMX_VIDEO_DEC_GET_CLASS (self);
 
   GST_DEBUG_OBJECT (self, "Handling frame");
 
-  GST_VIDEO_DECODER_STREAM_LOCK (self);
-  flush_flag = self->flush_flag;
-  GST_VIDEO_DECODER_STREAM_UNLOCK (self);
-
-  if (flush_flag) {
+  if (self->flush_flag) {
```
Comment (mmontero) [inline omx/gstomxvideodec.c]: keep protection

File: omx/gstomxvideodec.c [hunk 7]
```diff
@@ -2655,7 +2657,7 @@ gst_omx_video_dec_drain (GstVideoDecoder * decoder, gboolean is_eos)
 
   klass = GST_OMX_VIDEO_DEC_GET_CLASS (self);
 
-  if (!self->started) {
+  if (!self->started || self->flush_flag) {
```
Comment (mmontero) [inline omx/gstomxvideodec.c:2673]: protect flush_flag and started

**Review by mmontero — CHANGES_REQUESTED** [review]

File: omx/gstomxvideodec.c (hunk 4 above)
Comment (cgomezv) [reply omx/gstomxvideodec.c]: Added in 649bdbaf446bccb9760859a2c0f8ea3e5ad6f03c

File: omx/gstomxvideodec.c (hunk 5 above)
Comment (cgomezv) [reply omx/gstomxvideodec.c:1628]: It's already protected from line 1616

File: omx/gstomxvideodec.c (hunk 6 above)
Comment (cgomezv) [reply omx/gstomxvideodec.c]: Added in 649bdbaf446bccb9760859a2c0f8ea3e5ad6f03c

File: omx/gstomxvideodec.c (hunk 7 above)
Comment (cgomezv) [reply omx/gstomxvideodec.c:2673]: The thread is already locked, it gets unlocked in line 2683, and locked again when this function returns.

**Review by mmontero — APPROVED** [review]

File: omx/gstomxvideofilter.c [hunk 8]
```diff
@@ -362,6 +362,8 @@ gst_omx_video_filter_init (GstOMXVideoFilter * self,
   self->output_buffers = GST_OMX_VIDEO_FILTER_OUTPUT_BUFFERS_DEFAULT;
 
   g_rec_mutex_init (&self->stream_lock);
+  g_mutex_init (&self->filter_lock);
```
Comment (joselopez10) [inline omx/gstomxvideofilter.c:365]: Consider letting the reader know (briefly) why this is used. Something like: _The lock helps processing 1 buffer at a time, which in consequence prevents pipe deadlock during reverse playback_

**Review by joselopez10 — CHANGES_REQUESTED** [review]

File: omx/gstomxvideofilter.c (hunk 8 above)
Comment (cgomezv) [reply omx/gstomxvideofilter.c:365]: Comments added in b5747cd1914ada2395121f1cb91323dc2b665e84

**Review by joselopez10 — APPROVED** [review]

**Pull request merged**